File: extra samples/csharp/DemoButtonsHIDSerial/DemoButtonsForm.cs
/*
 DemoButtonsForm.cs

 Controlling program for the DemoButtons program which allow user to input a signature on an STU
 and reproduces it on a Window on the PC

 Copyright (c) 2015 Wacom GmbH. All rights reserved.

*/
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

namespace DemoButtons
{
    public partial class DemoButtonsForm : Form
    {
        public void print(string txt)
        {
            txtDisplay.Text += txt + "\r\n";
            txtDisplay.SelectionStart = txtDisplay.Text.Length; // scroll to end
            txtDisplay.ScrollToCaret();

        }

        public DemoButtonsForm()
        {
            InitializeComponent();
            radHID.Checked = true;
        }

        private void button1_Click(object sender, EventArgs e)
        {

            if (radHID.Checked)
            {
                captureSignatureHID();
            }
            else
            {
                captureSignatureSerial();
            }

        }

        private void captureSignatureHID()
        {
            wgssSTU.SerialInterface serialInterface = new wgssSTU.SerialInterface();
            wgssSTU.UsbDevices usbDevices = new wgssSTU.UsbDevices();

            if (usbDevices.Count != 0)
            {
                try
                {
                    wgssSTU.IUsbDevice usbDevice = usbDevices[0]; // select a device

                    SignatureForm demo = new SignatureForm(this, usbDevice, serialInterface, txtCOM.Text, txtBaudRate.Text, true);
                    DialogResult res = demo.ShowDialog();
                    //print("SignatureForm returned: " + res.ToString());
                    if (res == DialogResult.OK)
                    {
                        DisplaySignature(demo);
                    }
                
                    demo.Dispose();
                }
                catch (Exception ex)
                {
                    MessageBox.Show(ex.Message);
                }
            }
            else
            {
                MessageBox.Show("No STU devices attached");
            }
        }

        private void captureSignatureSerial()
        {
            int baudRate;
            string fileNameCOMPort;
            wgssSTU.IUsbDevice usbDevice;
            wgssSTU.SerialInterface serialInterface;

            usbDevice = null;
            serialInterface = new wgssSTU.SerialInterface();

            fileNameCOMPort = txtCOM.Text;
            baudRate = int.Parse(txtBaudRate.Text);

            try
            {
                SignatureForm demo = new SignatureForm(this, usbDevice, serialInterface, txtCOM.Text, txtBaudRate.Text, false);
                DialogResult res = demo.ShowDialog();
                print("SignatureForm returned: " + res.ToString());
                if (res == DialogResult.OK)
                {
                    DisplaySignature(demo);
                }
                demo.Dispose();
            }
            catch (Exception ex)
            {
                MessageBox.Show("No STU devices attached");
            }
        }

        private void radHID_CheckedChanged(object sender, EventArgs e)
        {
            radioSelectionCheck();
        }

        private void radSerial_CheckedChanged(object sender, EventArgs e)
        {
            radioSelectionCheck();
        }

        private void radioSelectionCheck()
        {
            if (radHID.Checked == true)
            {
                txtCOM.Enabled = false;
                txtBaudRate.Enabled = false;
            }
            else
            {
                txtCOM.Enabled = true;
                txtBaudRate.Enabled = true;
                txtBaudRate.Text = "128000";
                txtCOM.Focus();
            }
        }

        private void button2_Click(object sender, EventArgs e)
        {
            Application.Exit();
        }

        private void DisplaySignature(SignatureForm demo)
        {
            Bitmap bitmap;

            bitmap = demo.GetSigImage();
            // resize the image to fit the screen if needed
            int scale = 1;       
            if (bitmap.Width > 400)
                scale = 4;
            pictureBox1.Size = new Size(bitmap.Width / scale, bitmap.Height / scale);
            pictureBox1.SizeMode = PictureBoxSizeMode.StretchImage;
            pictureBox1.Image = bitmap;
            pictureBox1.Parent = this;
            //centre the image in the panel
            int x, y;
            x = panel1.Location.X + ((panel1.Width - pictureBox1.Width) / 2);
            y = panel1.Location.Y + ((panel1.Height - pictureBox1.Height) / 2);
            this.pictureBox1.Location = new Point(x, y);
            pictureBox1.BringToFront();
            //bitmap.Save("C:\\temp\\sig.png", System.Drawing.Imaging.ImageFormat.Png); // to save the image to disk
        }

        private void panel1_Paint(object sender, PaintEventArgs e)
        {

        }
    }
}

File: extra samples/vb/DemoButtonsHIDSerial/DemoButtonsForm.vb
'
' DemoButtonsForm.vb
'
' Allow user to input a signature
'
' Copyright (c) 2015 Wacom GmbH. All rights reserved.
'
'
Public Class DemoButtonsForm

    Private Sub DemoButtonsForm_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Load
      print("Ready")
      radHID.Checked = True
   End Sub

   Private Sub btnSign_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSign.Click

      If (radHID.Checked) Then
         captureSignatureHID()
      Else
         captureSignatureSerial()
      End If

   End Sub

   Private Sub captureSignatureHID()
      Dim usbDevices = New wgssSTU.UsbDevices()
      Dim serialInterface = New wgssSTU.SerialInterface()

      If (usbDevices.Count <> 0) Then
         Try
            print("Found " + usbDevices.Count.ToString() + " devices")
            Dim usbDevice = usbDevices(0)
            Dim demo = New SignatureForm(Me, usbDevice, serialInterface, txtCOM.Text, txtBaudRate.Text, True)
            Dim res = demo.ShowDialog()
            print("SignatureForm returned: " + res.ToString())
            If (res = DialogResult.OK) Then
               DisplaySignature(demo)
            End If

         Catch ex As Exception
            MsgBox("Exception: " + ex.Message)
         End Try

      Else
         print("No devices found")
      End If
   End Sub

   Private Sub captureSignatureSerial()

      Dim baudRate As Int32
      Dim fileNameCOMPort As String

      Dim usbDevice As wgssSTU.IUsbDevice
      Dim serialInterface = New wgssSTU.SerialInterface()

      fileNameCOMPort = txtCOM.Text
      baudRate = Int32.Parse(txtBaudRate.Text)

      Try
         Dim demo = New SignatureForm(Me, usbDevice, serialInterface, txtCOM.Text, txtBaudRate.Text, False)
         Dim Res As New DialogResult
         Res = demo.ShowDialog()
         print("SignatureForm returned: " + Res.ToString())
         If (Res = DialogResult.OK) Then
            DisplaySignature(demo)
            demo.Dispose()
         End If

      Catch ex As Exception
         MessageBox.Show("No STU devices attached: " + ex.Message)
      End Try

   End Sub

   Private Sub btnClose_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnClose.Click
        Me.Close()
    End Sub
    Public Sub print(ByVal txt As String)
        txtDisplay.Text += txt + vbCrLf
        txtDisplay.SelectionStart = txtDisplay.Text.Length ' scroll to end
        txtDisplay.ScrollToCaret()
    End Sub
    Private Sub DisplaySignature(ByVal demo As SignatureForm)
      Dim bitmap As Bitmap

      bitmap = demo.GetSigImage()
      ' resize the image to fit the screen
      Dim scale = 1       ' halve the image size if needed
      If bitmap.Width > 400 Then
         scale = 2
      End If

      PictureBox1.Size = New Size(bitmap.Width / scale, bitmap.Height / scale)
      PictureBox1.Image = bitmap
      'centre the image in the panel
      Dim x, y As Integer
      x = Panel1.Location.X + ((Panel1.Width - Panel1.Width) / 2)
      y = Panel1.Location.Y + ((Panel1.Height - Panel1.Height) / 2)
      Me.Panel1.Location = New Point(x, y)

      Panel1.BringToFront()

      'bitmap.Save("C:\\temp\\sig.png", Imaging.ImageFormat.Png) ' to save the image to disk
   End Sub

   Private Sub radHID_CheckedChanged(sender As Object, e As EventArgs) Handles radHID.CheckedChanged
      radioSelectionCheck()
   End Sub

   Private Sub radSerial_CheckedChanged(sender As Object, e As EventArgs) Handles radSerial.CheckedChanged
      radioSelectionCheck()
   End Sub

   Private Sub radioSelectionCheck()

      If (radHID.Checked = True) Then
         txtCOM.Enabled = False
         txtBaudRate.Enabled = False
      Else
         txtCOM.Enabled = True
         For Each serialPort As String In My.Computer.Ports.SerialPortNames
            txtCOM.Text = serialPort
         Next

         txtBaudRate.Enabled = True
         txtBaudRate.Text = "128000"
         txtCOM.Focus()
      End If
   End Sub
End Class
